core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala

```diff
@@ -72,7 +73,8 @@ final class ShuffleBlockFetcherIterator(
     maxBlocksInFlightPerAddress: Int,
     maxReqSizeShuffleToMem: Long,
     detectCorrupt: Boolean)
-  extends Iterator[(BlockId, InputStream)] with DownloadFileManager with Logging {
+  extends Iterator[(BlockId, InputStream)] with DownloadFileManager with TaskCompletionListener
```

Comment:
I don't it's a good idea for ShuffleBlockFetchIterator to be a subclass of TaskCompletionListener. 
What's wrong with original solution?

Reply:
The main reason is that TaskCompletionListener is added in one place (in `initialize` method) and needs to be removed in another one (in `cleanup` method).
![image](https://user-images.githubusercontent.com/1523889/48833554-abe64780-ed8c-11e8-9da0-ef826918a275.png)
Will introduce a field for TaskCompletionListener instead. Thank you!

Reply:
Another field sounds reasonable.

Reply:
Introduced [the corresponding field](https://github.com/apache/spark/pull/23083/files#diff-27109eb30a77542d377c936e0d134420R156).
```scala
  /**
   * Task completion callback to be called in both success as well as failure cases to cleanup.
   * It may not be called at all in case the `cleanup` method has already been called before
   * task completion.
   */
  private[this] val cleanupTaskCompletionListener = (_: TaskContext) => cleanup()
```